feat(gerrit): support moving changes to different target branches

felipecrs opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

This implements the feature of `updatePr()` which allows updating the target branch of a change. Very useful when the default branch name changes, for example.

This feature is supported for most platforms, but was not yet implemented for Gerrit.

During the implementation, I also realized that `getBranchPr()` was incorrectly implemented for Gerrit. In fact, it took me a while to understand that the `targetBranch` parameter should only be used to prioritize the search, not to filter the results. The only other platform which implemented it was Azure (https://github.com/renovatebot/renovate/discussions/23929). Therefore, to avoid future confusion, I added a proper JSDoc comment to the `targetBranch` parameter of `getBranchPr()`.

## Context

Please select one of the below:

- [x] This closes an existing Issue: https://github.com/renovatebot/renovate/discussions/39048
- [ ] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

<!-- We request this information to assist reviewers in identifying AI-generated errors and other issues specific to AI usage. While we typically permit the use of AI tools, we appreciate being notified when they are employed. -->

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [ ] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [x] Yes — substantive assistance (AI generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

The public repository: https://review.gerrithub.io/q/project:felipecrs/renovate-tutorial+status:open

```log
 INFO: Branch updated (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
       "commitSha": "b54864e04ca431933fbbf21d882f90d960450752"
DEBUG: Ensuring PR (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
DEBUG: There are 0 errors and 0 warnings (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
DEBUG: Found existing PR (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
DEBUG: Processing existing PR (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
DEBUG: PR base branch has changed (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
       "branchName": "renovate/commander-14.x",
       "oldBaseBranch": "main",
       "newBaseBranch": "new-main"
DEBUG: PR body changed (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
       "prTitle": "Update dependency commander to v14"
DEBUG: updatePr(1225670, Update dependency commander to v14) (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
 INFO: PR updated (repository=felipecrs/renovate-tutorial, baseBranch=new-main, branch=renovate/commander-14.x)
       "pr": 1225670,
       "prTitle": "Update dependency commander to v14"
```

<img width="1929" height="415" alt="image" src="https://github.com/user-attachments/assets/c9ca2215-ead7-4760-a6d8-7091a234edc6" />

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

RahulGautamSingh commented on lib/modules/platform/types.ts:
```diff
@@ -288,7 +288,18 @@ export interface Platform {
-  getBranchPr(branchName: string, targetBranch?: string): Promise<Pr | null>;
+  getBranchPr(
+    branchName: string,
+    /**
```
Also, mention that it is only being used for azure as of now. So it's displayed when someone tries to use it for another platform.

felipecrs commented on lib/modules/platform/gerrit/scm.spec.ts:
```diff
@@ -383,15 +382,65 @@ describe('modules/platform/gerrit/scm', () => {
+        }),
+      ).toBe('commitSha');
+      expect(git.prepareCommit).toHaveBeenCalledExactlyOnceWith({
+        baseBranch: 'new-main',
```
The reason why push happens with the older branch is explained [here](https://github.com/renovatebot/renovate/pull/39046/files/946f856e93ca203b692261026f2c0d4c85085477#diff-8afaa9cd83fcfe6ec06ac50ddba887c114873443b59f0e82725fceea2d9f51daR156-R160).

In short: if we push to the new branch, we will end up creating a new change. Therefore we push to the old branch so that we don't create a new change and then it can be moved in `updatePr()`.

felipecrs commented on lib/modules/platform/gerrit/index.ts:
```diff
@@ -174,6 +174,9 @@ export async function updatePr(prConfig: UpdatePrConfig): Promise<void> {
       TAG_PULL_REQUEST_BODY,
     );
   }
+  if (prConfig.targetBranch) {
```
I think the spirit of `updatePr()` is that it only gets called with stuff that needs to change, i.e. that check was already done before calling `updatePr()`, no?

But I can surely add this extra check.

felipecrs replied:
Information about the current branch is not available at this point, so I don't think that can be done without adding one more API call.

Instead, I'll check if I can add a protection in `moveChange` to avoid throwing in case the change is already in the wanted branch.

felipecrs commented on lib/modules/platform/gerrit/index.spec.ts:
```diff
@@ -356,72 +373,52 @@ describe('modules/platform/gerrit/index', () => {
-      ]);
     });
 
-    it('getBranchPr() - found even without targetBranch', async () => {
```
Actually never mind. I reconsidered and I think the best is to restore the tests. Done in https://github.com/renovatebot/renovate/pull/39046/commits/fc3da99f2d75434fc103ddfcb27d381a4bae1e67.

viceice commented on lib/modules/platform/types.ts:
```diff
@@ -288,7 +288,20 @@ export interface Platform {
-  getBranchPr(branchName: string, targetBranch?: string): Promise<Pr | null>;
+  getBranchPr(
+    branchName: string,
+    /**
```
move to tsdoc style comment above the function, so it's visible at call site in editor like vscode

viceice reviewed (commented):
otherwise LGTM

RahulGautamSingh reviewed (commented):
LGTM, except other comments.

felipecrs committed:
feat(gerrit): support moving changes to different target branches

Signed-off-by: Felipe Santos <felipe.santos@ericsson.com>

felipecrs commented:
@RahulGautamSingh, @viceice I had to update my git identity and therefore I squashed all commits from this PR into one.

There are no changes since last review round except a trivial merge conflict resolution. This PR is good to go from my side.

RahulGautamSingh reviewed (approved):
LGTM

viceice reviewed (commented):
otherwise LGTM

viceice committed:
Update lib/modules/platform/types.ts

felipecrs committed:
Merge branch 'main' into move-change
